### ISSUE TITLE
Architecture and Modularization of SIGTURK Library

### ISSUE THREAD

**Issue by mehmetoguzderin** [sigturk/sigturk#98]
Although we made robust & reasonable choices for programming and scripting of any code artifact we may manually or automatically produce with core guidelines (see https://sigturk.com/sigturk Code Guidelines), the discussion never went far into the architecture as we were to make prototype. Given the short amount of time, I propose that we should start with brainstorming collectively.

I would like to argue for a delicate balance of generality and specialization in consideration of historical trends of NLP, and how to specifically favor intaking data from linguistics for both processing of treatises related to Turkic and beyond, from reconstructions to attestations.

As a concrete point, I would like to exemplify with sequence of input. Obviously, we'll need to be containing than discriminating, hence, we can't have a single "chosen" way. But when doing this, an opportunity emerges: think of tokenization. Tokenization is seen as a separate task by norms, but if we make whether it is a tokenization a matter of metadata, we can have a general method named "transcribe" which takes a transcription method a, and converts to transcription method b, and it can have a shortest path to optimally go from one transcription to another, and treat tokenization as also a transcription method. Such an approach increases commonality by making a "normalize" shared. Imagine like the following diagram (just tentative sketch):

```mermaid
classDiagram
    class transcription {
        +Old Turkic
        +IPA
        +tiktoken_cl100k_base
    }
    class Old_Turkic {
        +transcribe()
    }
    class IPA {
        +transcribe()
    }
    class tiktoken_cl100k_base {
        +transcribe()
    }
    class normalization {
        +operate(transcription)
    }

    transcription --> Old_Turkic: contains
    transcription --> IPA: contains
    transcription --> tiktoken_cl100k_base: contains
    normalization --> transcription: operates on
```

Let's dialogue to make something that will increase productivity and have reasonable defaults.

**Comment by mehmetoguzderin** [comment]
Additional thought: any brainstorming on boundaries between transcription, transliteration, and translation for these blocks can have significant impact on trajectory

**Comment by mehmetoguzderin** [comment]
An inspiration source potentially: https://konlpy.org/en/latest/

**Comment by mehmetoguzderin** [comment]
From chat @j0ma : Spacy, Stanza

I also mentioned PyTorch Lightning

**Comment by mehmetoguzderin** [comment]
I was pondering a bit more and if we really want to simplify things by finding shortest path from one input to another output, not necessarily taking in pipeline from user, we might need to opt for a graph API like Taskflow or Tensorflow Graph mode. Given the eager mode's friendliness in REPL for notebooks it'd be critical to have a balance or be explicit about two ways coexisting.

For defaults such as transcribing or translating an input to another for better performance in pipeline, we could also use RLHF for productive defaults (initial scoring maybe only from officers team, no need to try to go to scale fttb imo).

```
session = ...;
task = session.add_taskgraph()
task.add_input(ty=Old_Uyghur_Script)
task.add_output(ty=Sentiment_Analysis)
session.solve(task)
task(...) // outputs some sort of sentiment analysis through any n-steps that might be needed to have good or valid results. steps should be printable at any time for reproduction purposes
```

**Comment by mehmetoguzderin** [comment]
Thinking a tiny more about design process (as we need to ramp things up), I think best way might be to just do TDD, test-driven development. Let's discuss on Monday

**Comment by mehmetoguzderin** [comment]
We've thought that components are necessary because ecosystem is evolving really fast, @j0ma particularly pointed out one example we might want to abstract: https://github.com/isi-nlp/uroman

**Comment by mehmetoguzderin** [comment]
It seems like tabular data structuring from papers or other papers is an applicable method. We should probably have tabularize or extract_table methods in our graph system. From presentation of @defnecirci

**Comment by mehmetoguzderin** [comment]
We've also thought of a pipeline taking open-access publications from different languages (Turkish, Russian, Chinese, etc.) and then structuring them through the library
